### DISCUSSION TITLE
Output Post-Middleware

### DISCUSSION THREAD

**Discussion by is-it-ayush** [trpc/trpc#3770] [Q&A]
Is there a way to achieve a `.output()` but instead of defining at the end of each procedure. I want this to act as a output middleware. I know that it's possible to
 ```ts
const myOutputMiddleware = t.middleware(async ({ ctx, next }) => {
  const result = await next();
  if (result.ok && result.data) {
    console.log(result.data);
  }
  return result;
});
``` 
However the problem is that to wrap my data returned from procedure into a specific type. Say, the procedure `example.getAll` returns a list of users along with a event name defined inside the procedure (string). Now I want to wrap the `data` inside a object of type,
```ts
{
  event: data.event_name,
  result: data.users,
  timestamp: 219421491,
  path: ctx.path,
  type: ctx.type,
  ... // etc. you get the idea
}
```
I want to save this object & return the `data.users` to the client. Now, I could achieve this using `.output()` but the problem is I want this globally for all proceduresm (like a middleware) or a router. This becomes a hassle if you have 100+ procedures. Any way to solve this?

**Comment by KATT** [comment]
Do you need the output to be validated as well?

**Reply by is-it-ayush** [reply]
If it's possible then that would be great Alex!

**Reply by KATT** [reply]
You can probably do some `AsyncLocalStorage`-magic where you have middlewares like:

1. Call `opts.next()` with a middleware in a asynclocalstorage
1. Have an output validator
1. Add output to local storage


Something like

```ts
procedure
  .use((opts) => { 
      // run opts.next() in a asynclocalstorage
   })
  .output(z.object({ 
    //
  }))
  .use(opts => {
     // attach `opts.next().data` in localstorage
  })
  .query(() => '...')
```

**Reply by is-it-ayush** [reply]
So essentially the solution for now is to go for the same post-middlware method but store my returned data in a local store? I've heard that AsyncLocalStorage comes with a performance overhead. So, adding that as an local output cache on every procedure doesn't seem a good option. I've also dug into the trpc codebase a little bit & played a bit with the internal functions but couldn't find a way. Are there any better ways possible supposing this "output middleware" will be used in almost every procedure? Thank You though!

**Reply by KATT** [reply]
Maybe create a procedure factory function with the output as an arg


Sorry, this is a niche use-case and it's very time-consuming to help you here, you gotta experiment on your own and if it needs some feature in tRPC, please make a feature request

**Reply by is-it-ayush** [reply]
Completely Understandable! Thank you for the help Katt. I appreciate it. I'll experiment with a factory function as you suggested. I'll update you if I find something worthy enough to be a feature in trpc. Have a good day! 🙌
